### PR TITLE
Instrument middleware processing

### PR DESCRIPTION
**Pull request by dasch**
Adds ActiveSupport::Notifications instrumentation of the processing of each middleware in the stack.

@jeremy here's the first PR!

### CODE REVIEW

**Comment by dasch** [comment]
?r @jeremy

**Comment by dasch** [comment]
I think the failing test is unrelated to my change...

**Review by jeremy — COMMENTED** [review]
LGTM. @tenderlove any concerns about stack depth?

**Comment by dasch** [comment]
@jeremy @tenderlove mind if I bump this?

File: actionpack/lib/action_dispatch/middleware/stack.rb [hunk 1]
```diff
@@ -34,7 +34,28 @@ def inspect
...
       def build(app)
-        klass.new(app, *args, &block)
+        InstrumentationProxy.new(klass.new(app, *args, &block), inspect)
+      end
+    end
+
+    # This class is used to instrument the execution of a single middleware.
+    # It proxies the `call` method transparently and instruments the method
+    # call.
+    class InstrumentationProxy
```
Comment (jeremy) [inline actionpack/lib/action_dispatch/middleware/stack.rb]: Will these proxy objects show up in any user-facing API? Concerned whether we'll be obscuring the middleware stack anywhere. (If so, we could do something like generate proxy subclasses named after the middleware class.)

File: actionpack/lib/action_dispatch/middleware/stack.rb [hunk 2]
```diff
@@ -34,7 +34,28 @@ def inspect
...
+      EVENT_NAME = "process_middleware.action_dispatch"
```
Comment (jeremy) [inline actionpack/lib/action_dispatch/middleware/stack.rb]: We can use the string literal directly; no need for a constant.

File: actionpack/test/dispatch/middleware_stack_test.rb [hunk 3]
```diff
@@ -109,6 +120,24 @@ def test_delete_works
...
+    end
+
+    ActiveSupport::Notifications.subscribed(subscriber, "process_middleware.action_dispatch") do
+      app.call(env)
+    end
+
+    assert_equal 2, events.count
+    assert_equal ["MiddlewareStackTest::BarMiddleware", "MiddlewareStackTest::FooMiddleware"], events.map { |e| e.payload[:middleware] }
+  end
+
```
Comment (jeremy) [inline actionpack/test/dispatch/middleware_stack_test.rb]: 👌🏼

**Review by jeremy — APPROVED** [review]

File: actionpack/lib/action_dispatch/middleware/stack.rb (hunk 1 above)
Comment (dasch) [reply actionpack/lib/action_dispatch/middleware/stack.rb]: We're already wrapping the middleware instances in Rails, I believe.

File: actionpack/lib/action_dispatch/middleware/stack.rb (hunk 2 above)
Comment (dasch) [reply actionpack/lib/action_dispatch/middleware/stack.rb]: It was done more for code clarity purposes...

**Comment by jhawthorn** [comment]
I think this PR broke the actionpack test suite

https://travis-ci.org/rails/rails/jobs/479883440#L1733

```
Failure:
ShowExceptions::ShowFailsafeExceptionsTest#test_render_failsafe_exception [test/controller/show_exceptions_test.rb:108]:
Expected: "text/plain"
  Actual: "text/html"
```

**Comment by jhawthorn** [comment]
I think this PR broke the actionview test suite

https://travis-ci.org/rails/rails/jobs/479883440#L1733

```
ShowExceptions::ShowFailsafeExceptionsTest#test_render_failsafe_exception [/home/travis/build/rails/rails/actionpack/test/controller/show_exceptions_test.rb:108]:
Expected: "text/plain"
  Actual: "text/html"
```

I was super sure it was us in #35669 😅, so much that I deleted this comment the first time I posted to double check, but reverting this does fix the test.

**Comment by tenderlove** [comment]
> LGTM. @tenderlove any concerns about stack depth?

Yes, I have concerns.  I think this is going to add a bunch of overhead.  Why do we need this exactly?  Profiling?  If it's for profiling, I strongly suggest we better integrate an actual profiling tool like `StackProf`.

**Comment by dasch** [comment]
It’s for production monitoring, mostly. This will be used for tracing.

**Pull request merged**